### PR TITLE
Move providing responsabilities from bitswap to blockservice

### PR DESCRIPTION
**Pull request by gammazero**
- bitswap/server: remove provide
- blockservice: add session workaround to work with wrapped blockservices
- blockservice: add WithProvider option
- blockservice: remove session embeding in context

Replaces #534 by @Jorropo which was outdated enough to make merging difficult.

### CODE REVIEW

**Comment by gammazero** [comment]
Needs kubo PR to test

**Comment by hsanjuan** [comment]
Reviewing this and probably merging into #641 ...

File: blockservice/blockservice.go [hunk 1]
```diff
@@ -133,23 +157,13 @@ func (s *blockService) Allowlist() verifcid.Allowlist {
...
-// session will be created. Otherwise, the current exchange will be used
-// directly.
-// Sessions are lazily setup, this is cheap.
-func NewSession(ctx context.Context, bs BlockService) *Session {
-	ses := grabSessionFromContext(ctx, bs)
-	if ses != nil {
-		return ses
-	}
-
-	return newSession(ctx, bs)
```
Comment (hsanjuan) [inline blockservice/blockservice.go:148]: I'm not sure why session-grabbing from context is removed... I remember old discussions about embedding sessions in contexts etc. but not sure if used...

**Comment by hsanjuan** [comment]
This was more or less made part of #641

**Pull request closed without merging**